<?xml version="1.0"?>
<robot name="2cyl">
    <link name="world"/>
        <link name="base_link">
            <inertial>
                <origin xyz="0.0 0.0 0.05" rpy="0.0 0.0 0.0"/>
                <mass value="1.0"/>
                <inertia ixx="27" ixy="0.0" ixz="0.0" iyy="0.005" iyz="0.0" izz="0.0027"/>
            </inertial>
            <visual>
                <origin xyz="0.0 0.0 0.05" rpy="0.0 0.0 0.0"/>
                <geometry>
                    <cylinder radius="1.6" length="0.1"/>
                </geometry>
                <material name="red">
                    <color rgba="1.0 0.0 0.0 1.0"/>
                </material>
            </visual>
            <collision>
                <origin xyz="0.0 0.0 0.05" rpy="0.0 0.0 0.0"/>
                <geometry>
                    <cylinder radius="1.6" length="0.1"/>
                </geometry>
            </collision>
        </link>

    <joint name="joint" type="fixed">
        <parent link="world"/>
        <child link="base_link"/>
    </joint>



    


    <link name="link1">
        <inertial>
            <origin xyz="0.0 0.0 0.1" rpy="0.0 0.0 0.0"/>
            <mass value="1.0"/>
            <inertia ixx="27.0" ixy="0.0" ixz="0.0" iyy="0.0027" iyz="0.0" izz="0.005"/>
        </inertial>
        <visual>
            <origin xyz="0.0 0.0 0.1" rpy="0.0 0.0 0.0"/>
            <geometry>
                <cylinder radius="0.1" length="0.2"/>
            </geometry>
            <material name="blue">
                <color rgba="0.0 0.0 1.0 1.0"/>
            </material>
        </visual>
        <collision>
            <origin xyz="0.0 0.0 0.1" rpy="0.0 0.0 0.0"/>
            <geometry>
                <cylinder radius="0.1" length="0.2"/>
            </geometry>
        </collision>
    </link>

    <joint name="joint1" type="continuous">
      <parent link="base_link"/>
      <child link="link1"/>
      <origin xyz="0.0 0 0.2" rpy="0.0 0.0 0.0"/>
      <axis xyz="0.0 0.0 1.0"/>
    </joint>



    <link name="link2">
        <inertial>
            <origin xyz="0.0 0.0 0.4" rpy="0.0 0.0 0.0"/>
            <mass value="1.0"/>
            <inertia ixx="27.0" ixy="0.0" ixz="0.0" iyy="0.0027" iyz="0.0" izz="0.005"/>
        </inertial>
        <visual>
            <origin xyz="0.0 0.0 0.4" rpy="0.0 0.0 0.0"/>
            <geometry>
                <cylinder radius="0.1" length="0.8"/>
            </geometry>
            <material name="blue">
                <color rgba="0.0 0.0 1.0 1.0"/>
            </material>
        </visual>
        <collision>
            <origin xyz="0.0 0.0 0.4" rpy="0.0 0.0 0.0"/>
            <geometry>
                <cylinder radius="0.1" length="0.8"/>
            </geometry>
        </collision>
    </link>


    <joint name="joint2" type="revolute">
        <origin xyz="-0.2 0.0 0.15" rpy="0.0 0.0 0.0"/>
        <parent link="link1"/>
        <child link="link2"/>
        <axis xyz="1.0 0.0 0.0"/>
        <limit lower="0.8" upper="-0.8" effort="0.0" velocity="0.0"/> 
    </joint>



    <link name="link3">
        <inertial>
            <origin xyz="0.0 0.0 0.4" rpy="0.0 0.0 0.0"/>
            <mass value="1.0"/>
            <inertia ixx="27.0" ixy="0.0" ixz="0.0" iyy="0.0027" iyz="0.0" izz="0.005"/>
        </inertial>
        <visual>
            <origin xyz="0.0 0.0 0.4" rpy="0.0 0.0 0.0"/>
            <geometry>
                <cylinder radius="0.1" length="0.8"/>
            </geometry>
            <material name="blue">
                <color rgba="0.0 0.0 1.0 1.0"/>
            </material>
        </visual>
        <collision>
            <origin xyz="0.0 0.0 0.4" rpy="0.0 0.0 0.0"/>
            <geometry>
                <cylinder radius="0.1" length="0.8"/>
            </geometry>
        </collision>
    </link>


    <joint name="joint3" type="revolute">
        <origin xyz="0.2 0.0 0.75" rpy="0.0 0.0 0.0"/>
        <parent link="link2"/>
        <child link="link3"/>
        <axis xyz="1.0 0.0 0.0"/>
        <limit lower="0.7" upper="-0.7" effort="0.0" velocity="0.0"/>
    </joint>




    <link name="link4">
        <inertial>
            <origin xyz="0.0 0.0 0.4" rpy="0.0 0.0 0.0"/>
            <mass value="1.0"/>
            <inertia ixx="27.0" ixy="0.0" ixz="0.0" iyy="0.0027" iyz="0.0" izz="0.005"/>
        </inertial>
        <visual>
            <origin xyz="0.0 0.0 0.4" rpy="0.0 0.0 0.0"/>
            <geometry>
                <cylinder radius="0.1" length="0.8"/>
            </geometry>
            <material name="blue">
                <color rgba="0.0 0.0 1.0 1.0"/>
            </material>
        </visual>
        <collision>
            <origin xyz="0.0 0.0 0.4" rpy="0.0 0.0 0.0"/>
            <geometry>
                <cylinder radius="0.1" length="0.8"/>
            </geometry>
        </collision>
    </link>


    <joint name="joint4" type="revolute">
        <origin xyz="-0.2 0.0 0.75" rpy="0.0 0.0 0.0"/>
        <parent link="link3"/>
        <child link="link4"/>
        <axis xyz="1.0 0.0 0.0"/>
        <limit lower="0.7" upper="-0.7" effort="0.0" velocity="0.0"/> 
    </joint>





    <link name="link5">
        <inertial>
            <origin xyz="0.0 0.0 0.125" rpy="0.0 0.0 0.0"/>
            <mass value="0.5"/>
            <inertia ixx="0.0" ixy="0.0" ixz="0.0" iyy="0.0" iyz="0.0" izz="0.0"/>
        </inertial>
        <visual>
            <origin xyz="0.0 0.0 0.125" rpy="0.0 0.0 0.0"/>
            <geometry>
                <box size="0.25 0.25 0.25"/>
            </geometry>
            <material name="blue">
                <color rgba="0.0 0.0 1.0 1.0"/>
            </material>
        </visual>
        <collision>
            <origin xyz="0.0 0.0 0.125" rpy="0.0 0.0 0.0"/>
            <geometry>
                <box size="0.25 0.25 0.25"/>
            </geometry>
        </collision>
    </link>


    <joint name="joint5" type="fixed">
        <origin xyz="0.0 0.0 0.8" rpy="0.0 0.0 0.0"/>
        <parent link="link4"/>
        <child link="link5"/>
       <!-- <axis xyz="0.0 0.0 0.0"/>
        <limit lower="0.0" upper="0.0" effort="0.0" velocity="0.0"/>-->
    </joint>






    <!--<link name="link4">
        <inertial>
            <origin xyz="0.0 0.0 0.3" rpy="0.0 0.0 0.0"/>
            <mass value="0.25"/>
            <inertia ixx="27.0" ixy="0.0" ixz="0.0" iyy="0.0027" iyz="0.0" izz="0.005"/>
        </inertial>
        <visual name="">
            <origin xyz="0.0 0.0 0.3" rpy="0.0 0.0 0.0"/>
            <geometry>
                <cylinder radius="0.1" length="0.6"/>
            </geometry>
            <material name="">
                <color rgba="1.0 0.0 0.0 1.0"/>
            </material>
        </visual>
        <collision>
            <origin xyz="0.0 0.0 0.3" rpy="0.0 0.0 0.0"/>
            <geometry>
                <cylinder radius="0.1" length="0.6"/>
            </geometry>
        </collision>
    </link>


    <joint name="joint4" type="continuous">
        <origin xyz="-0.2 0.0 0.75" rpy="0.0 0.0 0.0"/>
        <parent link="link3"/>
        <child link="link4"/>
        <axis xyz="1.0 0.0 0.0"/>
        <limit lower="0.0" upper="0.0" effort="0.0" velocity="0.0"/>
    </joint>




    <link name="link5">
        <inertial>
            <origin xyz="-0.05 0.0 0.1" rpy="0.0 0.0 0.0"/>
            <mass value="0.25"/>
            <inertia ixx="27.0" ixy="0.0" ixz="0.0" iyy="0.0027" iyz="0.0" izz="0.005"/>
        </inertial>
        <visual name="">
            <origin xyz="-0.05 0.0 0.1" rpy="0.0 0.0 0.0"/>
            <geometry>
                <cylinder radius="0.05" length="0.2"/>
            </geometry>
            <material name="">
                <color rgba="1.0 0.0 0.0 1.0"/>
            </material>
        </visual>
        <collision>
            <origin xyz="-0.05 0.0 0.1" rpy="0.0 0.0 0.0"/>
            <geometry>
                <cylinder radius="0.05" length="0.2"/>
            </geometry>
        </collision>
    </link>


    <joint name="joint5" type="continuous">
        <origin xyz="-0.1 0.0 0.8" rpy="0.0 0.0 0.0"/>
        <parent link="link3"/>
        <child link="link5"/>
        <axis xyz="0.0 1.0 0.0"/>
        <limit lower="0.0" upper="0.0" effort="0.0" velocity="0.0"/>
    </joint>





    <transmission name="trans5">
       <type>transmission_interface/SimpleTransmission</type>
       <actuator name="motor5">
         <mechanicalReduction>1</mechanicalReduction>
       </actuator>
       <joint name="joint5">
         <hardwareInterface>PositionJointInterface</hardwareInterface>
       </joint>
    </transmission>  




    <transmission name="trans4">
       <type>transmission_interface/SimpleTransmission</type>
       <actuator name="motor4">
         <mechanicalReduction>1</mechanicalReduction>
       </actuator>
       <joint name="joint4">
         <hardwareInterface>PositionJointInterface</hardwareInterface>
       </joint>
    </transmission> -->

    


    <transmission name="trans1">
       <type>transmission_interface/SimpleTransmission</type>
       <actuator name="motor1">
         <mechanicalReduction>1</mechanicalReduction>
       </actuator>
       <joint name="joint1">
         <hardwareInterface>PositionJointInterface</hardwareInterface>
       </joint>
    </transmission>


    <transmission name="trans2">
       <type>transmission_interface/SimpleTransmission</type>
       <actuator name="motor2">
         <mechanicalReduction>1</mechanicalReduction>
       </actuator>
       <joint name="joint2">
         <hardwareInterface>PositionJointInterface</hardwareInterface>
       </joint>
    </transmission>



    <transmission name="trans3">
       <type>transmission_interface/SimpleTransmission</type>
       <actuator name="motor3">
         <mechanicalReduction>1</mechanicalReduction>
       </actuator>
       <joint name="joint3">
         <hardwareInterface>PositionJointInterface</hardwareInterface>
       </joint>
    </transmission>



    <transmission name="trans4">
       <type>transmission_interface/SimpleTransmission</type>
       <actuator name="motor4">
         <mechanicalReduction>1</mechanicalReduction>
       </actuator>
       <joint name="joint4">
         <hardwareInterface>PositionJointInterface</hardwareInterface>
       </joint>
    </transmission>


    

    

    <gazebo>
       <plugin name="gazebo_ros_control" filename="libgazebo_ros_control.so">
         <robotNamespace>/</robotNamespace>
       </plugin>
    </gazebo>

    <gazebo>
      <plugin name="joint_state_publisher" filename="libgazebo_ros_joint_state_publisher">
        <jointName>joint1</jointName>
      </plugin>  
    </gazebo>
</robot>